content/blog/2019/11/2019-11-21-welcome-to-the-matrix.adoc

```diff
@@ -0,0 +1,389 @@
+    axes { ... }
+    excludes { ... }
+    agent {
+        label "${PLATFORM}-agent"
+    }
+    stages { ... }
+    // ...
+}
+----
+
+
+Occasionally I run my pipeline manually from the Jenkins Web UI. 
+When I do that, I'd like to be able to select just one platform to run, but I can't use `excludes` for this.
+The `axis` and `exclude` directives define the static set of cells that make up the matrix. 
+That set of combinations is generated before the start of the run, before any parameters. 
+
+So, I can't add or remove cells from a matrix after the job has started but I can use a "per-cell" directives. 
+I'll add a `choice` parameter with the list of platforms and use a `when` directive inside `matrix` to control which cells in the matrix are run.
+The log below shows a portion of the output when I select `PLATFORM_FILTER = 'mac'`.
+
+[source, groovy]
+----
+pipeline {
+    parameters {
+        choice(name: 'PLATFORM_FILTER', choices: ['all', 'linux', 'windows', 'mac'], description: 'Run on specific platform')
+    }
+    agent none
+    stages {
+        stage('BuildAndTest') {
+            matrix {
+                agent {
+                    label "${PLATFORM}-agent"
+                }
+                when { anyOf {
+                    expression { params.PLATFORM_FILTER == 'all' }
+                    expression { params.PLATFORM_FILTER == env.PLATFORM }
+                } }
+                axes {
+                    axis {
+                        name 'PLATFORM'
+                        values 'linux', 'windows', 'mac'
+                    }
+                    axis {
+                        name 'BROWSER'
+                        values 'firefox', 'chrome', 'safari', 'edge'
+                    }
+                }
+                excludes {
+                    exclude {
+                        axis {
+                            name 'PLATFORM'
+                            values 'linux'
+                        }
+                        axis {
+                            name 'BROWSER'
+                            values 'safari'
+                        }
+                    }
+                    exclude {
+                        axis {
+                            name 'PLATFORM'
+                            notValues 'windows'
+                        }
+                        axis {
+                            name 'BROWSER'
+                            values 'edge'
+                        }
+                    }
+                }
+                stages {
+                    stage('Build') {
+                        steps {
+                            echo "Do Build for ${PLATFORM} - ${BROWSER}"
+                        }
+                    }
+                    stage('Test') {
+                        steps {
+                            echo "Do Test for ${PLATFORM} - ${BROWSER}"
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+----
+
```

Comment:
A sentence or two explaining what you just did here, between the pipeline and the log output, would be useful. "When I ran the job, I set the `PLATFORM_FILTER` parameter to run only those cells for which `PLATFORM` is set to `mac`.

Reply:
@kshultzCB I don't like to put single sentences between code blocks like this as I think they get lost... I'll give it a try.